### ISSUE TITLE
Show assembly side-by-side like rust.godbolt.org does

### ISSUE THREAD

**Issue by gnzlbg** [integer32llc/rust-playground#210]
I really like how https://rust.godbolt.org/ shows the assembly generated side-by-side, removes the "cruft" of the assembly output, and uses colors to represent from which code which parts of the assembly were generated.

The play ground should do this for assembly too. It would also be great if it could do this for MIR and LLVM IR as well.

**Comment by shepmaster** [comment]
> side-by-side

Could you expand on how the existing side-by-side display would need to be changed to be equivalent?



> removes the "cruft" of the assembly output, and uses colors to represent from which code which parts of the assembly were generated.

Do you know how this is done?

> if it could do this for MIR and LLVM IR as well.

Do you have a suggestion on how do to this?

**Comment by gnzlbg** [comment]
> Could you expand on how the existing side-by-side display would need to be changed to be equivalent?

The current display is code top, output bottom (or am I missing something):

![screen-shot-2017-09-11-at-16 19 38](https://user-images.githubusercontent.com/904614/30279372-43a08090-970d-11e7-80ab-b28f4efc5dc4.png)

This is "ok" when running a program. I think it would be better if the warnings and the program output would be put on the right side next to the code when the display is big enough. 

Anyhow, when looking at MIR, LLVM IR, or ASM it is very tedious to scroll down the bottom window, skipping the warnings, and then trying to find out relationships between the emitted code and the code I wrote. 

This is how godbolt does that:

![screen-shot-2017-09-11-at-16 23 51](https://user-images.githubusercontent.com/904614/30279894-b6254906-970e-11e7-97f2-27a429bf67a4.png)

I like the pane on the right way better. One can use the same setup with multiple "tabs", but with a pane on the right instead of under the code, and switch from output, to warnings, to asm, mir, llvm ir. I like that the warnings are not shown every time, because when I am looking at the ASM I don't care about them. I also like how the colors are use to make it clear where the assembly originates from. With the current playground, I need to read all the assembly and constantly thin "is this what I am looking for" till I arrive at the relevant piece of information. With godbolt, I just look at the code, check the color, and look right and find the relevant assembly. I find this pure genius.

> Do you know how this is done?

No idea, but godbolt is opensource under the BSD license: https://github.com/mattgodbolt/compiler-explorer .

> Do you have a suggestion on how do to this?

Nope, but if this information is somehow preserved till the ASM is generated, so it must be possible to emit it for LLVM IR and MIR as well.

**Comment by shepmaster** [comment]
> or am I missing something):

Make your window wider, or:

![image](https://user-images.githubusercontent.com/174509/30281360-29b96ed0-96e0-11e7-9ac5-71f22a22b09c.png)

**Comment by gnzlbg** [comment]
@shepmaster Indeed, my window was pretty wide (IMO, Chrome 60.0.3112.101 64-bit), but if I make it a little wider it switch to a vertical pane. Sorry for the noise, I did not know this.

![screen shot 2017-09-11 at 17 00 28](https://user-images.githubusercontent.com/904614/30281502-e4c5ef46-9712-11e7-8f85-bb6c9f9944a8.png)

**Comment by shepmaster** [comment]
> my window was pretty wide (IMO, Chrome 60.0.3112.101 64-bit), but if I make it a little wider it does automatically switch to a vertical pane

Yep; I used the old playground for a long time and never knew it did that; someone with a bigger screen had to tell me about it. This version at least allows you to override it and force it to one way or the other regardless of screensize.

**Comment by RReverser** [comment]
It would be still nice to have colored markers for asm. Ideally, I'd also love if playground, just like godbolt.org, could show the compiled assembly as you type instead of having to press a button each time, but I don't know if that would be considered too heavy for the playground server?

**Comment by gnzlbg** [comment]
@shepmaster 

Maybe I could extend [`cargo-asm`](https://github.com/gnzlbg/cargo-asm) to generate json output or I could refactor it into a library+binary so that the playground could use it.

**Comment by gnzlbg** [comment]
@chandlore could you point me to where that script and the front-ends part of it are? 

On the `cargo-asm` side of things I could add an option to `cargo-asm` that spits output in a way that is easy for the front-end to consume, for example, a json array of assembly lines with associated metadata: to which Rust code file does the assembly line correspond to, to which line within the file, maybe a string of the rust line, what type of assembly node ist it: a directive, a label (with label name), a comment, an instruction: which type of instruction  (instruction name, a jump-type instruction to a label, a call instruction, ...).

On the playground side of things we could invoke cargo-asm for the file and the `--json` flag, parse the json and do something cool with it:
  - add options to choose what to show: raw (all of the assembly, with directives, comments, irrelevant labels, ...), concise (leaving directive, comments, irrelevant labels, ... out)
  - use bg colors like godbolt to map rust code to assembly code, on hover of the assembly code/rust code, the color associated with the line hovered should become "intense"
  - format assembly instructions: give the instruction names a different color, make them bold, hyperlink them to the reference (e.g. [http://www.felixcloutier.com/x86/{instruction_name}.xml](http://www.felixcloutier.com/x86/XSAVE.xml)), show a summary of what the instruction does on hover, ...
  - format labels: detect from the json the jump instructions and format their label argument and labels alike, hyperlink them so that clicking on a label in an instruction jumps the assembly view to the label
  - figure out what to do for assembly instructions that correspond to code not in the file that the user wrote (e.g. because its on the std library, inlined): we could add it as a comment to the rust file in the places where it belongs and link that from the assembly view to there, we could also link the code to the src file (e.g. in the std library docs)

Emitting json from cargo-asm is the easy part. Doing something cool with it is where the meat of the work will be. Depending on what we want to do with it I can emit more or less json. All the meta-data pieces mentioned above are part of the cargo-asm AST, so if I just serialize that to json all of that should be available. 

If there is something cool that we want to do that needs some more metadata, as long as generating that meta-data does not require a connection to the internet (e.g. linking instructions to the reference and providing summaries) cargo-asm can probably generate it with minor modifications. Anything that requires an internet connection is probably best done in the front-end.

**Comment by shepmaster** [comment]
Does cargo-asm work on stable? Otherwise, it doesn't seem like a viable solution for the playground.

**Comment by chandlore** [comment]
@gnzlbg You can find the current filter script at [asm_cleanup.rs#L29-L133](https://github.com/integer32llc/rust-playground/blob/8829e92e5698cecfe09d9f5e683160e1b4f6844b/ui/src/asm_cleanup.rs#L29-L133).  It does not hook into the front end, instead it runs as [part of the compile function](https://github.com/integer32llc/rust-playground/blob/8829e92e5698cecfe09d9f5e683160e1b4f6844b/ui/src/sandbox.rs#L150-L171) of the sandbox module on the server. 

I like all the ideas you listed above, and nice work on ```cargo-asm``` 😃.  It does fail to compile on stable ```rustc 1.23.0 (766bd11c8 2018-01-01)``` for me due to ```#![feature(match_default_bindings)]```, but I hope it can be made to work with the playground.

**Comment by gnzlbg** [comment]
Cargo-asm should work for stable rust programs (otherwise it is a bug), but
it requires a nightly compiler to build. The only unstable feature used is
in cargo-asm itself is match_default_bindings and that is merely an
ergonomic improvement (so this should be trivial to fix).
